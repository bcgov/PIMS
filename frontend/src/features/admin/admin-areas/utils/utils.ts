import { IAdministrativeArea, IApiAdminArea } from '../interfaces';

/** generate administrative area ready for use in api endpoints */
export const toApiAdminArea = (adminArea: IAdministrativeArea, name?: string): IApiAdminArea => {
  return {
    name: name ?? adminArea.name,
    boundaryType: adminArea.boundaryType ?? undefined,
    abbreviation: adminArea.abbreviation ?? undefined,
    groupName: adminArea.groupName ?? undefined,
    id: adminArea.id,
    isDisabled: adminArea.isDiabled ?? false,
    code: String(adminArea.id),
    rowVersion: adminArea.rowVersion ?? '',
    type: 'AdministrativaArea',
  };
};
