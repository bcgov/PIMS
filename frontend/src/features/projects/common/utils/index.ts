export * from './getNumber';
export * from './calcNetProceeds';
export * from './calcGainBeforeSpl';
