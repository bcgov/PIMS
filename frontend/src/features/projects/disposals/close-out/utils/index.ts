export * from './calcGainBeforeSpl';
export * from './calcNetProceeds';
export * from './getNumber';
