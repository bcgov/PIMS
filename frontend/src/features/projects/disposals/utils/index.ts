export * from './convertToPropertyType';
export * from './convertToPropertyTypeName';
export * from './getMostRecentEvaluation';
export * from './getMostRecentFiscal';
export * from './toForm';
export * from './toModel';
export * from './toPropertyForm';
export * from './toPropertyModel';
