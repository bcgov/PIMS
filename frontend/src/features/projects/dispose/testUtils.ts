import {
  IProjectTask,
  DisposeWorkflowStatus,
  IProject,
  ReviewWorkflowStatus,
  ITask,
  SPPApprovalTabs,
  NoteTypes,
} from '../common';
import { ProjectActions } from 'constants/actionTypes';
import * as reducerTypes from 'constants/reducerTypes';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as API from 'constants/API';

export const mockTasks: IProjectTask[] = [
  {
    projectNumber: 123,
    taskId: 1,
    isOptional: true,
    isCompleted: false,
    name: 'task-0',
    description: 'one',
    taskType: 1,
    sortOrder: 0,
    completedOn: new Date(),
    statusId: 0,
    statusCode: DisposeWorkflowStatus.RequiredDocumentation,
  },
  {
    projectNumber: 123,
    taskId: 2,
    isOptional: true,
    isCompleted: true,
    name: 'task-1',
    description: 'two',
    taskType: 1,
    sortOrder: 0,
    completedOn: new Date(),
    statusId: 0,
    statusCode: DisposeWorkflowStatus.RequiredDocumentation,
  },
];

export const mockProject: IProject = {
  projectNumber: 'test-01',
  name: 'my project',
  description: 'my project description',
  privateNote: 'private note',
  publicNote: 'public note',
  notes: [
    {
      noteType: NoteTypes.General,
      note: 'general',
    },
    {
      noteType: NoteTypes.Public,
      note: 'public',
    },
    {
      noteType: NoteTypes.Private,
      note: 'private',
    },
    {
      noteType: NoteTypes.Exemption,
      note: 'exemption',
    },
    {
      noteType: NoteTypes.AgencyInterest,
      note: 'agencyinterest',
    },
    {
      noteType: NoteTypes.Financial,
      note: 'financial',
    },
    {
      noteType: NoteTypes.PreMarketing,
      note: 'premarketing',
    },
    {
      noteType: NoteTypes.Marketing,
      note: 'marketing',
    },
    {
      noteType: NoteTypes.ContractInPlace,
      note: 'contractinplace',
    },
    {
      noteType: NoteTypes.Reporting,
      note: 'reporting',
    },
    {
      noteType: NoteTypes.LoanTerms,
      note: 'loanterms',
    },
    {
      noteType: NoteTypes.Adjustment,
      note: 'adjustment',
    },
    {
      noteType: NoteTypes.SppCost,
      note: 'sppcost',
    },
    {
      noteType: NoteTypes.SppGain,
      note: 'sppgain',
    },
    {
      noteType: NoteTypes.SalesHistory,
      note: 'saleshistory',
    },
    {
      noteType: NoteTypes.CloseOut,
      note: 'closeout',
    },
    {
      noteType: NoteTypes.Comments,
      note: 'comments',
    },
  ],
  properties: [],
  agencyId: 1,
  statusId: 0,
  statusCode: ReviewWorkflowStatus.ApprovedForErp,
  status: {
    id: 1,
    name: 'Approved for ERP',
    sortOrder: 0,
    route: '',
    description: '',
    workflowCode: '',
    code: ReviewWorkflowStatus.ApprovedForErp,
    isMilestone: true,
    tasks: [],
    isOptional: false,
    isActive: true,
  },
  tierLevelId: 1,
  tasks: mockTasks,
  note: 'my notes',
  id: 1,
  fiscalYear: 2020,
  projectAgencyResponses: [],
};

export const tasks: ITask[] = [
  {
    taskId: 1,
    name: 'task-0',
    sortOrder: 0,
    description: 'test',
    taskType: 1,
  },
  {
    taskId: 2,
    name: 'task-1',
    sortOrder: 0,
    description: 'test',
    taskType: 1,
  },
];

export const mockWorkflow = [
  {
    description:
      'A new draft project that is not ready to submit to apply to be added to the Surplus Property Program.',
    route: '/projects/draft',
    isMilestone: false,
    code: 'DR',
    id: 1,
    name: 'Draft',
    isDisabled: false,
    sortOrder: 0,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUM=',
    isActive: true,
    tasks: [],
  },
  {
    description: 'Add properties to the project.',
    route: '/projects/properties',
    isMilestone: false,
    code: 'DR-P',
    id: 2,
    name: 'Select Properties',
    isDisabled: false,
    sortOrder: 1,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUQ=',
    isActive: true,
    tasks: [],
  },
  {
    description: 'Assign tier level, classification and update current financial information.',
    route: '/projects/information',
    isMilestone: false,
    code: 'DR-I',
    id: 3,
    name: 'Update Information',
    isDisabled: false,
    sortOrder: 2,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUU=',
    isActive: true,
    tasks: [],
  },
  {
    description:
      'Required documentation has been completed and sent (Surplus Declaration \u0026 Readiness Checklist, Triple Bottom Line).',
    route: '/projects/documentation',
    isMilestone: false,
    code: 'DR-D',
    id: 4,
    name: 'Required Documentation',
    isDisabled: false,
    sortOrder: 3,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUY=',
    isActive: true,
    tasks: [],
  },
  {
    description: 'The project is ready to be approved by owning agency.',
    route: '/projects/approval',
    isMilestone: false,
    code: 'DR-A',
    id: 5,
    name: 'Approval',
    isDisabled: false,
    sortOrder: 4,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUc=',
    isActive: true,
    tasks: [],
  },
  {
    description:
      'The project has been submitted for review to be added to the Surplus Property Program.',
    route: '/projects/review',
    isMilestone: false,
    code: 'DR-RE',
    id: 6,
    name: 'Review',
    isDisabled: false,
    sortOrder: 5,
    isOptional: false,
    workflowCode: 'SUBMIT-DISPOSAL',
    type: 'ProjectStatus',
    createdOn: '2020-06-29T04:46:45.2666667',
    rowVersion: 'AAAAAAAACUg=',
    isActive: true,
    tasks: [],
  },
];

const mockStore = configureMockStore([thunk]);
export const getStore = (mockProject: IProject, tab?: SPPApprovalTabs) =>
  mockStore({
    [reducerTypes.LOOKUP_CODE]: {
      lookupCodes: [
        {
          code: 'BCT',
          name: 'BC Transit',
          id: 1,
          type: API.AGENCY_CODE_SET_NAME,
        },
      ],
    },
    [reducerTypes.ProjectReducers.PROJECT]: { project: mockProject },
    [reducerTypes.ProjectReducers.TASKS]: tasks,
    [reducerTypes.NETWORK]: {
      [ProjectActions.GET_PROJECT]: {},
    },
    [reducerTypes.ProjectReducers.SPL_TAB]: tab,
  });
