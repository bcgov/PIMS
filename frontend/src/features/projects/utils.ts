import { DisposalWorkflows } from './constants';
import { IProject } from './interfaces';

/**
 * Determine if the specified 'project' has been in the SPL workflow.
 * @param project The project object.
 * @returns True if the project history shows SPL.
 */
export const wasInSpl = (project: IProject) => {
  return project?.statusHistory.some((s) => s.workflow === DisposalWorkflows.Spl) ?? false;
};

/**
 * Determine if the specified 'project' has been in the ERP workflow.
 * @param project The project object.
 * @returns True if the project history shows ERP.
 */
export const wasInErp = (project: IProject) => {
  return project?.statusHistory.some((s) => s.workflow === DisposalWorkflows.Erp) ?? false;
};
