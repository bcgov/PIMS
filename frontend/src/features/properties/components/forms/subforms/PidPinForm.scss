.pid-pin-form {
  align-items: center;
  margin-bottom: 1em !important;

  .left-column {
    text-align: end;
  }

  .left-column.pin-column {
    margin-left: 1.2em;
  }

  .input-small {
    width: 12em;
  }

  .no-padding {
    padding: 0 0 0 0.3em;
  }
}
