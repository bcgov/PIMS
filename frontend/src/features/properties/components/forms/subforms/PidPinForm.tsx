import { FunctionComponent } from 'react';
import React from 'react';
import { Input, Form } from 'components/common/form';
import { PidPinTooltip } from '../strings';
import { Label } from 'components/common/Label';

interface PidPinProps {
  handlePidChange: (pid: string, nameSpace?: string) => void;
  handlePinChange: (pin: string, nameSpace?: string) => void;
  nameSpace?: string;
  disabled?: boolean;
}
export const defaultPidPinFormValues: {
  pid: number | '';
  pin: number | '';
} = {
  pid: '',
  pin: '',
};

/**
 * The pidFormatter is used to format the specified PID value
 * @param {string} pid This is the target PID to be formatted
 */
export const pidFormatter = (pid: string) => {
  pid = pid.padStart(9, '0');
  const regex = /(\d\d\d)[\s-]?(\d\d\d)[\s-]?(\d\d\d)/;
  const format = pid.match(regex);
  if (format !== null && format.length === 4) {
    pid = `${format[1]}-${format[2]}-${format[3]}`;
  }
  return pid;
};

const PidPinForm: FunctionComponent<PidPinProps> = (props: PidPinProps) => {
  const withNameSpace: Function = (fieldName: string) => {
    const { nameSpace } = props;
    return nameSpace ? `${nameSpace}.${fieldName}` : fieldName;
  };

  return (
    <>
      <Form.Row className="flex-nowrap pid-pin">
        <Label>PID</Label>
        <Input
          required={true}
          displayErrorTooltips
          className="input-small"
          disabled={props.disabled}
          pattern={RegExp(/^[\d\- ]*$/)}
          onBlurFormatter={(pid: string) => {
            if (pid?.length > 0) {
              return pid.replace(pid, pidFormatter(pid));
            }
            return '';
          }}
          field={withNameSpace('pid')}
        />
        <Label style={{ paddingLeft: '5px' }}>PIN</Label>
        <Input
          required={true}
          displayErrorTooltips
          className="input-small"
          tooltip={PidPinTooltip}
          disabled={props.disabled}
          field={withNameSpace('pin')}
          onBlurFormatter={(pin: number) => {
            if (pin > 0) {
              return pin;
            }
            return '';
          }}
          type="number"
        />
      </Form.Row>
    </>
  );
};

export default PidPinForm;
