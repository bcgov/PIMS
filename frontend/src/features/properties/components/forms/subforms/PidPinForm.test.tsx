import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react';
import PidPinForm from './PidPinForm';
import useKeycloakWrapper, { IKeycloak } from 'hooks/useKeycloakWrapper';
import { Formik, Form } from 'formik';
import { fillInput } from 'utils/testUtils';
import { IParcel } from 'actions/parcelsActions';

jest.mock('lodash/debounce', () => jest.fn(fn => fn));
jest.mock('hooks/useKeycloakWrapper');
(useKeycloakWrapper as jest.Mock<Partial<IKeycloak>>).mockReturnValue({ hasClaim: () => true });

describe('PidPin sub-form', () => {
  let form: JSX.Element;
  let handlePidChange = jest.fn();
  let handlePinChange = jest.fn();
  beforeEach(() => {
    form = (
      <Formik<Partial<IParcel>>
        initialValues={{ pid: '', pin: '', projectNumber: '' }}
        initialTouched={{ pid: true, pin: true }}
        validateOnChange={false}
        onSubmit={() => {}}
      >
        {formikProps => (
          <Form>
            <PidPinForm handlePidChange={handlePidChange} handlePinChange={handlePinChange} />
          </Form>
        )}
      </Formik>
    );
  });
  afterEach(() => {
    handlePinChange.mockReset();
    handlePidChange.mockReset();
    cleanup();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = renderer.create(form).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('formats PID values', async () => {
    const { container } = render(form);
    const { input } = await fillInput(container, 'pid', '123456789');
    expect(input).toHaveValue('123-456-789');
  });
});
