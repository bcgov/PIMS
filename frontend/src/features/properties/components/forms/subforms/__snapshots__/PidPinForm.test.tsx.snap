// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PidPin sub-form renders correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="flex-nowrap pid-pin form-row"
  >
    <p
      className="label"
    >
      PID
    </p>
    <div
      className="required form-group"
    >
      <input
        className="input-small form-control"
        id="input-pid"
        name="pid"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        value=""
      />
    </div>
    <p
      className="label"
      style={
        Object {
          "paddingLeft": "5px",
        }
      }
    >
      PIN
    </p>
    <div
      className="required form-group"
    >
      <input
        className="input-small form-control"
        id="input-pin"
        name="pin"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
      <svg
        className="tooltip-icon"
        fill="currentColor"
        height="1em"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
        />
      </svg>
    </div>
  </div>
</form>
`;
